### PR TITLE
Add Thycotic DevOps Secrets Vault lookup plugin

### PR DESCRIPTION
**Pull request by amigus**
##### SUMMARY
Add a lookup plug-in capable of retrieving Secrets from Thycotic's DevOps Secrets Vault. The plug-in uses python-dsv-sdk which is available here: https://pypi.org/project/python-dsv-sdk/

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
dsv

##### ADDITIONAL INFORMATION
- This is a re-submission of a closed PR with the changes suggested in that PR.
- The cyclomatic complexity of this plug-in is 5 and the SDK on which it's based is 3.  As such, I've opted not to include a pytest.  If future versions have increased complexity, I will add tests at that point.

### CODE REVIEW

**Comment by felixfontein** [comment]
Link to the original PR: #86

**Comment by amigus** [comment]
Yeah, sorry folks, but I screwed up the original PR so I had to resubmit.  FYI, the changes suggested on PR #86 are incorporated.

**Comment by amigus** [comment]
It looks to me like the failures are due to the absence of tests.

**Comment by amigus** [comment]
@Akasurde this is the plugin with your changes applied.

File: plugins/lookup/dsv.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+RETURN = r"""
+_list:
+    description:
+        - One or more JSON responses to `GET /secrets/{path}`
+        - See https://dsv.thycotic.com/api/index.html#operation/getSecret"""
+
+EXAMPLES = r"""
+- hosts: localhost
+  vars:
+      secret: "{{ lookup('dsv', '/test/secret') }}"
```
Comment (felixfontein) [inline plugins/lookup/dsv.py:73]:
```suggestion
      secret: "{{ lookup('community.general.dsv', '/test/secret') }}"
```

File: plugins/lookup/dsv.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+  tasks:
+      - debug: msg="the password is {{ secret["data"]["password"] }}"
```
Comment (felixfontein) [inline plugins/lookup/dsv.py:75]: Please use YAML syntax:
```suggestion
      - debug:
          msg: 'the password is {{ secret["data"]["password"] }}'
```

File: plugins/lookup/dsv.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+                if path == "":
+                    raise AnsibleOptionsError("Invalid secret path: %s" % term)
+
+                display.vvv(u"DevOps Secrets Vault GET /secrets/%s" % path)
+                result.append(vault.get_secret_json(path))
+            except SecretsVaultError as error:
+                raise AnsibleError(
+                    "DevOps Secrets Vault lookup failure: %s" % error.message
+                )
+        display.debug(u"dsv_lookup result: %s" % result)
```
Comment (felixfontein) [inline plugins/lookup/dsv.py:122]: Don't output potentially sensitive results to the log!
```suggestion
```

File: plugins/lookup/dsv.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+from ansible.errors import AnsibleError, AnsibleOptionsError
+
+try:
+    from thycotic.secrets.vault import (
+        SecretsVault,
+        SecretsVaultError,
+    )
+except ImportError:
+    raise AnsibleError("python-dsv-sdk must be installed to use this plugin")
```
Comment (felixfontein) [inline plugins/lookup/dsv.py:86]: This is not acceptable. Do what other lookup plugins do, and on failed import set a flag and only fail in the actual plugin code.

Comment (amigus) [reply plugins/lookup/dsv.py:87]: OK, so, to be clear about what you're asking for, I can keep the `except ImportError` but when it's triggered, just set a flag then test for the flag and raise AnsibleError from within `LookupModule.run()` correct?

Comment (amigus) [reply plugins/lookup/dsv.py:87]: Yeah, I found an example that works as described and I will copy that...

File: plugins/lookup/dsv.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        description: the first format parameter in the default `url_template`
+        env:
+            - name: DSV_TENANT
+        ini:
+            - section: dsv_lookup
+              key: tenant
+        required: True
+    tld:
+        default: com
+        description: the top-level domain of the tenant; the second format
```
Comment (felixfontein) [inline plugins/lookup/dsv.py:29]:
```suggestion
        description: The top-level domain of the tenant; the second format
```

File: plugins/lookup/dsv.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+            parameter in the default `url_template`
```
Comment (felixfontein) [inline plugins/lookup/dsv.py:30]:
```suggestion
            parameter in the default I(url_template).
```

File: plugins/lookup/dsv.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        env:
+            - name: DSV_TLD
+        ini:
+            - section: dsv_lookup
+              key: tld
+        required: False
+    client_id:
+        description: the client_id with which to request the Access Grant
```
Comment (felixfontein) [inline plugins/lookup/dsv.py:38]:
```suggestion
        description: The client_id with which to request the Access Grant.
```

File: plugins/lookup/dsv.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+        env:
+            - name: DSV_CLIENT_ID
+        ini:
+            - section: dsv_lookup
+              key: client_id
+        required: True
+    client_secret:
+        description: the client_secret associated with the specific client_id
```
Comment (felixfontein) [inline plugins/lookup/dsv.py:46]:
```suggestion
        description: The client secret associated with the specific I(client_id).
```

File: plugins/lookup/dsv.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+        env:
+            - name: DSV_CLIENT_SECRET
+        ini:
+            - section: dsv_lookup
+              key: client_secret
+        required: True
+    url_template:
+        default: https://{}.secretsvaultcloud.{}/v1
+        description: the path to append to the base URL to form a valid REST
```
Comment (felixfontein) [inline plugins/lookup/dsv.py:55]:
```suggestion
        description: The path to prepend to the base URL to form a valid REST
```
You don't do something like `/baseurl/https://{}.secretsvaultcloud.{}/v1`, right? :)

File: plugins/lookup/dsv.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+            API request
```
Comment (felixfontein) [inline plugins/lookup/dsv.py:56]:
```suggestion
            API request.
```

File: plugins/lookup/dsv.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+        env:
+            - name: DSV_URL_TEMPLATE
+        ini:
+            - section: dsv_lookup
+              key: url_template
+        required: False"""
```
Comment (felixfontein) [inline plugins/lookup/dsv.py:62]:
```suggestion
        required: False
"""
```
Makes diff nicer when adding new options.

File: plugins/lookup/dsv.py [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+
+RETURN = r"""
+_list:
+    description:
+        - One or more JSON responses to `GET /secrets/{path}`
+        - See https://dsv.thycotic.com/api/index.html#operation/getSecret"""
```
Comment (felixfontein) [inline plugins/lookup/dsv.py:68]:
```suggestion
        - See U(https://dsv.thycotic.com/api/index.html#operation/getSecret).
"""
```

File: plugins/lookup/dsv.py [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+            - name: DSV_URL_TEMPLATE
+        ini:
+            - section: dsv_lookup
+              key: url_template
+        required: False"""
+
+RETURN = r"""
+_list:
+    description:
+        - One or more JSON responses to `GET /secrets/{path}`
```
Comment (felixfontein) [inline plugins/lookup/dsv.py:67]:
```suggestion
        - One or more JSON responses to C(GET /secrets/{path}).
```

**Comment by gundalow** [comment]
@amigus Hi, anything we can do to help?
If you go to https://github.com/ansible-collections/community.general/pull/90/files you should be able to `add suggestion to batch` then ina single commit get all those review comments applied. After that locally you'll need to do `git pull origin master`

As a side note, when you make future PRs please remember to create a branch first. ie

```
git pull upstream master
git checkout -b myfeature
# do stuff
git commit
git push origin myfeature
```

File: plugins/lookup/dsv.py (hunk 9 above)
Comment (amigus) [reply plugins/lookup/dsv.py:55]: Um, yeah, that's totally how it works!  😋

**Comment by amigus** [comment]
OK, that's done.

RE: integration tests, we discussed it on IRC and agreed that my laziness was acceptable this time as the complexity of the code is so low and I promise to add integration testing along with any future changes.

**Comment by felixfontein** [comment]
@amigus I'm pretty sure I said that even if complexity is low, there *must* be tests. We accepted too many modules without tests in the past, and even if they started simple enough, they usually never got tests later on and eventually we had problems with them due to missing tests.

It should be really easy to add some unit tests which test the basic behavior of this plugin (same for the other lookup).

**Comment by amigus** [comment]
I've added a unit test.

File: plugins/lookup/dsv.py [hunk 14]
```diff
@@ -0,0 +1,138 @@
+# -*- coding: utf-8 -*-
+# Copyright: (c) 2020, Adam Migus <adam@migus.org>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+# python 3 headers, required if submitting to Ansible
```
Comment (felixfontein) [inline plugins/lookup/dsv.py:4]: Please remove this line.
```suggestion

```

File: plugins/lookup/dsv.py [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+DOCUMENTATION = r"""
+lookup: dsv
+author: Adam Migus (adam@migus.org)
+short_description: Get secrets from Thycotic DevOps Secrets Vault
```
Comment (felixfontein) [inline plugins/lookup/dsv.py:12]:
```suggestion
short_description: Get secrets from Thycotic DevOps Secrets Vault
version_added: 1.0.0
```

File: plugins/lookup/dsv.py [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+description:
+    - Uses the Thycotic DevOps Secrets Vault Python SDK to get Secrets from a
+       DSV `tenant` using a `client_id` and `client_secret`.
```
Comment (felixfontein) [inline plugins/lookup/dsv.py:15]:
```suggestion
      DSV I(tenant) using a I(client_id) and I(client_secret).
```

File: plugins/lookup/dsv.py [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+requirements:
+    - python-dsv-sdk - https://pypi.org/project/python-dsv-sdk/
+options:
+    _terms:
+        description: The path to the secret, e.g. C(/staging/servers/web1).
+        required: True
```
Comment (felixfontein) [inline plugins/lookup/dsv.py:21]:
```suggestion
        required: true
```
Use either `true`/`false` everywhere, or `yes`/`no` everywhere.

File: tests/unit/plugins/lookup/test_dsv.py [hunk 18]
```diff
@@ -0,0 +1,57 @@
+# -*- coding: utf-8 -*-
+# (c) 2020, Adam Migus <adam@migus.org>
+#
+# This file is part of Ansible
```
Comment (felixfontein) [inline tests/unit/plugins/lookup/test_dsv.py:4]: Please use the short GPL header for this (`GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)` in one line below the copyright).

File: tests/unit/plugins/lookup/test_dsv.py [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+from ansible_collections.community.general.tests.unit.compat.mock import (
+    patch,
+    MagicMock,
+)
+from ansible_collections.community.general.plugins.lookup import dsv
+from ansible.plugins.loader import lookup_loader
+
+
+class MockSecretsVault(MagicMock):
+    RESPONSE = '{"foo":"bar"}'
```
Comment (felixfontein) [inline tests/unit/plugins/lookup/test_dsv.py:34]:
```suggestion
    RESPONSE = '{"foo": "bar"}'
```
See the failing sanity tests: https://app.shippable.com/github/ansible-collections/community.general/runs/2503/1/tests

File: plugins/lookup/dsv.py [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+        description: The path to the secret, e.g. C(/staging/servers/web1).
+        required: true
+    tenant:
+        description: The first format parameter in the default I(url_template).
+        env:
+            - name: DSV_TENANT
+        ini:
+            - section: dsv_lookup
+              key: tenant
+        required: True
```
Comment (felixfontein) [inline plugins/lookup/dsv.py:29]:
```suggestion
        required: true
```

File: plugins/lookup/dsv.py [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+    tld:
+        default: com
+        description: The top-level domain of the tenant; the second format
+            parameter in the default I(url_template).
+        env:
+            - name: DSV_TLD
+        ini:
+            - section: dsv_lookup
+              key: tld
+        required: False
```
Comment (felixfontein) [inline plugins/lookup/dsv.py:39]:
```suggestion
        required: false
```

File: plugins/lookup/dsv.py [hunk 22]
```diff
@@ -0,0 +1,138 @@
...
+    client_id:
+        description: The client_id with which to request the Access Grant.
+        env:
+            - name: DSV_CLIENT_ID
+        ini:
+            - section: dsv_lookup
+              key: client_id
+        required: True
```
Comment (felixfontein) [inline plugins/lookup/dsv.py:47]:
```suggestion
        required: true
```

File: plugins/lookup/dsv.py [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+    client_secret:
+        description: The client secret associated with the specific I(client_id).
+        env:
+            - name: DSV_CLIENT_SECRET
+        ini:
+            - section: dsv_lookup
+              key: client_secret
+        required: True
```
Comment (felixfontein) [inline plugins/lookup/dsv.py:55]:
```suggestion
        required: true
```

File: plugins/lookup/dsv.py [hunk 24]
```diff
@@ -0,0 +1,138 @@
...
+    url_template:
+        default: https://{}.secretsvaultcloud.{}/v1
+        description: The path to prepend to the base URL to form a valid REST
+            API request.
+        env:
+            - name: DSV_URL_TEMPLATE
+        ini:
+            - section: dsv_lookup
+              key: url_template
+        required: False
```
Comment (felixfontein) [inline plugins/lookup/dsv.py:65]:
```suggestion
        required: false
```

**Comment by felixfontein** [comment]
@amigus thanks for contributing this plugin with tests!

**Comment by Andersson007** [comment]
@amigus hey, could you please create a new pr to change `author: Adam Migus (adam@migus.org)` => `author: Adam Migus (@amigus)`
If you don't have time, let me know, i'll add it myself

**Comment by Andersson007** [comment]
the same https://github.com/ansible-collections/community.general/pull/91/files there. You could create one PR for both the plugins

**Comment by felixfontein** [comment]
@Andersson007 already fixed in #753.

**Comment by Andersson007** [comment]
@felixfontein ok, thanks for the info

**Pull request merged**